From 5bbc9ddba9d13cd2742f847ca05c594a872dca36 Mon Sep 17 00:00:00 2001
From: Sri Vidya Gunturi <sgunturi@qti.qualcomm.com>
Date: Thu, 17 Apr 2025 17:55:57 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL: hostapd: Modify the minimum length of
 action frame to 26

The action frames of length less than 27 are being dropped in the handle_action
API. The header length of Action frame is 24. And the minimum data length is 3.
Modify this length check to 26 as EPCS Teardown Action frame length is 26,
where the data length is 2.

As per the IEEE P802.11be/D7.0 Spec - 9.6.38.7 EPCS Priority Access Teardown
frame details, the EPCS Teardown Action frame data has 2 fields:
Action Category (37) and Action code - Teardown (5).

Signed-off-by: Sri Vidya Gunturi <sgunturi@qti.qualcomm.com>
---
 src/ap/hostapd.h    |  3 +++
 src/ap/ieee802_11.c |  2 +-
 src/ap/utils.c      | 23 +++++++++++++++++++++++
 3 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index b45c42b..2761f51 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -1068,4 +1068,7 @@ int hostapd_epcs_handle_cli(struct hostapd_data *hapd, char *pos,
 			    char *buf, size_t buflen);
 
 #endif /* CONFIG_IEEE80211BE */
+bool
+hostapd_verify_action_frame_has_min_length(const struct ieee80211_mgmt *mgmt,
+					   size_t len);
 #endif /* HOSTAPD_H */
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 6bd52a0..8c78098 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -6720,7 +6720,7 @@ static int handle_action(struct hostapd_data *hapd,
 	struct sta_info *sta;
 	u8 *action __maybe_unused;
 
-	if (len < IEEE80211_HDRLEN + 2 + 1) {
+	if (!hostapd_verify_action_frame_has_min_length(mgmt, len)) {
 		hostapd_logger(hapd, mgmt->sa, HOSTAPD_MODULE_IEEE80211,
 			       HOSTAPD_LEVEL_DEBUG,
 			       "handle_action - too short payload (len=%lu)",
diff --git a/src/ap/utils.c b/src/ap/utils.c
index e93e531..a5ba23e 100644
--- a/src/ap/utils.c
+++ b/src/ap/utils.c
@@ -110,3 +110,26 @@ void hostapd_prune_associations(struct hostapd_data *hapd, const u8 *addr,
 		hapd->iface->interfaces->for_each_interface(
 			hapd->iface->interfaces, prune_associations, &data);
 }
+
+bool
+hostapd_verify_action_frame_has_min_length(const struct ieee80211_mgmt *mgmt,
+					   size_t len)
+{
+	size_t min_len;
+
+	/* Ignore frames that are too short to even contain the frame category
+	 * and action code fields
+	 */
+	if (len < IEEE80211_HDRLEN + 2)
+		return false;
+
+	min_len = (mgmt->u.action.u.epcs_teardown.action ==
+		   WLAN_PROT_EHT_EPCS_ENABLE_TEARDOWN)
+		  ? IEEE80211_HDRLEN + 2
+		  : IEEE80211_HDRLEN + 3;
+
+	if (len < min_len)
+		return false;
+
+	return true;
+}
-- 
2.34.1

